src/syncback/snapshot.rs

```diff
@@ -0,0 +1,226 @@
+        new_ref: Ref,
+        old_ref: Option<Ref>,
+    ) -> anyhow::Result<Self> {
+        let mut snapshot = Self {
+            data: self.data,
+            old: old_ref,
+            new: new_ref,
+            path: PathBuf::new(),
+            middleware: None,
+        };
+        let middleware = get_best_middleware(&snapshot);
+        let name = name_for_inst(middleware, snapshot.new_inst(), snapshot.old_inst())?;
+        snapshot.path = base_path.join(name.as_ref());
+
+        Ok(snapshot)
+    }
+
+    /// Constructs a SyncbackSnapshot with the provided path and refs while
+    /// inheriting the data of the this snapshot.
+    #[inline]
+    pub fn with_new_path(&self, path: PathBuf, new_ref: Ref, old_ref: Option<Ref>) -> Self {
+        Self {
+            data: self.data,
+            old: old_ref,
+            new: new_ref,
+            path,
+            middleware: None,
+        }
+    }
+
+    /// Allows a middleware to be 'forced' onto a SyncbackSnapshot to override
+    /// the attempts to derive it.
+    #[inline]
+    pub fn middleware(mut self, middleware: Middleware) -> Self {
+        self.middleware = Some(middleware);
+        self
+    }
+
+    /// Returns a map of properties for an Instance from the 'new' tree
+    /// with filtering done to avoid noise. This method filters out properties
+    /// that are not meant to be present in Instances that are represented
+    /// specially by a path, like `LocalScript.Source` and `StringValue.Value`.
+    ///
+    /// This method is not necessary or desired for blobs like Rbxm or non-path
+    /// middlewares like JsonModel.
+    #[inline]
+    #[must_use]
+    pub fn get_path_filtered_properties(
+        &self,
+        new_ref: Ref,
+    ) -> Option<HashMap<&'sync str, &'sync Variant>> {
+        let inst = self.get_new_instance(new_ref)?;
+
+        // The only filtering we have to do is filter out properties that are
+        // special-cased in some capacity.
+        let properties = filter_properties(self.data.project, inst)
+            .into_iter()
+            .filter(|(name, _)| !filter_out_property(inst, name))
+            .collect();
+
+        Some(properties)
+    }
+
+    /// Returns a path to the provided Instance in the new DOM. This path is
+    /// where you would look for the object in Roblox Studio.
+    #[inline]
+    pub fn get_new_inst_path(&self, referent: Ref) -> String {
+        let mut path = Vec::new();
+        let mut path_capacity = 0;
+
+        let mut inst = self.get_new_instance(referent);
+        while let Some(instance) = inst {
+            path.push(&instance.name);
+            path_capacity += instance.name.len() + 1;
+            inst = self.get_new_instance(instance.parent());
+        }
+        let mut str = String::with_capacity(path_capacity);
+        while let Some(segment) = path.pop() {
+            str.push_str(segment);
+            str.push('/')
+        }
+        str.pop();
+
+        str
+    }
```

Comment:
As an aside, we can simplify this by using some standard library methods:
 ```suggestion
     pub fn get_new_inst_path(&self, referent: Ref) -> String {
        let mut path = Vec::new();

        let mut inst = self.get_new_instance(referent);
        while let Some(instance) = inst {
            path.push(instance.name.as_str());
            inst = self.get_new_instance(instance.parent());
        }

        path.reverse();
        path.join("/")
    }
```